service.cc

```diff
@@ -0,0 +1,348 @@
+#include "service.h"
+
+#include <chrono>
+#include <thread>
+#include <mutex>
+#include <vector>
+
+#include <google/protobuf/util/time_util.h>
+#include <google/protobuf/repeated_field.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+#include <grpcpp/server_context.h>
+#include <grpcpp/security/server_credentials.h>
+
+#include "service.grpc.pb.h"
+#include "data_storage_types.grpc.pb.h"
+
+grpc::Status Service::registeruser(grpc::ServerContext *context, const chirp::RegisterRequest *request, chirp::RegisterReply *reply) {
+  // check ifuser is already registered
+  std::string user_serialized = backend_client_.SendGetRequest(request->username());
+  if(!user_serialized.empty()) {
+    return grpc::Status(grpc::StatusCode::ALREADY_EXISTS, "User Already Exists");
+  }
+
+  // User doesn't exist, so save to backend
+  chirp::User user;
+  user.set_username(request->username());
+  user.SerializeToString(&user_serialized); 
+  backend_client_.SendPutRequest(request->username(), user_serialized);
+  return grpc::Status::OK;
+}
+
+grpc::Status Service::chirp(grpc::ServerContext *context, const chirp::ChirpRequest *request, chirp::ChirpReply *reply) {
+  std::lock_guard<std::mutex> guard(mutex_); // get lock
+
+  // check if user who is chirping exists in key value store
+  std::string user_exists = backend_client_.SendGetRequest(request->username());
+  if(user_exists.empty()) {
+    return grpc::Status(grpc::StatusCode::NOT_FOUND, "User not found");
+  }
+
+  // check for parent id's existance
+  if(!request->parent_id().empty()) {
+    std::string parent_exists = backend_client_.SendGetRequest(request->parent_id());
+    if(parent_exists.empty()) {
+      return grpc::Status(grpc::StatusCode::NOT_FOUND, "Parent ID not found");
+    }
+  }
+
+  // User does exist, so allow them to chirp
+  // Create new chirp ID based on current number of chirps
+  std::string num_chirps_str = backend_client_.SendGetRequest("num_chirps");
+  int num_chirps;
```

Comment:
Could num_chirps be initialized to 0 in the KVS in the constructor? Instead of having to create the key when you add the first chirp.

Reply:
yea I can do that